HostMirror: make return types consistent and simplify HostSpace selection

janciesko opened the pull request:
This PR refines Kokkos::Impl::HostMirror to return a “host mirror” type that matches the category of the template argument (execution space vs memory space vs device), and updates ViewTraits and unit tests to use that behavior consistently. It also simplifies and documents the HostSpace decision logic.

Motivation: Several call sites effectively need “the host mirror space of this thing,” but HostMirror behavior was inconsistent depending on whether you passed an execution space, memory space, or device. This change makes the function predictable and reduces surprising type mismatches across backends.

Changes:

* HostMirror API refinement: Introduces/uses `HostMirror<S>::Device`, `execution_space`, `memory_space`, and `Space` so the returned type aligns with the kind of S.
* Clarifies and simplifies `HostSpace selection logic`
* ViewTraits consistency: `host_mirror_space` is now derived consistently via `HostMirror<MemorySpace>` (rather than mixing `Space`, `nested ::memory_space`, etc.).
* Tests updated

janciesko committed:
Make HostMirror return-types depend on template arg
Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Make HostMirror return-types depend on template arg
Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Make ViewsTraits use HostMirror type consistently
Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Make ViewsTraits use HostMirror type consistently
Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

romintomasetti commented on core/src/Kokkos_View.hpp:
```diff
@@ -227,8 +227,7 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
-           Device<DefaultHostExecutionSpace,
-                  typename traits::host_mirror_space::memory_space>,
+           typename Impl::HostMirror<memory_space>::Device,
            typename traits::hooks_policy>;
```
This is not directly related to your PR but I would like to gather opinions here.

Why would the `host_mirror_type` not inherit from the memory trait ?

My first thoughts are:
- For `Kokkos::Unmanaged`, we should be cautious. For instance, is it meaningful that `create_host_mirror_view_and_copy` returns an unmanaged view ?
- For `Kokkos::Atomic`, do we expect that the `hhost_mirror_type` be atomic too ?

janciesko replied:
The host mirror used in the `View` template argument needs to be a `Device` (note the line that we're replacing) hence we construct it here from `memory_space` via `HostMirror`. 
You mean `create_mirror_view_and_copy`?

janciesko commented:
In CI checks I am seeing repeatedly: 
```
In file included from /home/runner/_work/kokkos/kokkos/kokkos/build/core/unit_test/hip/TestHIPManaged_ViewMapping_a.cpp:2:
/home/runner/_work/kokkos/kokkos/kokkos/core/unit_test/TestViewMapping_a.hpp:1041:17: error: reference to __host__ function 'use_count' in __host__ __device__ function
          if (a.use_count() != 3) ++e;
```
Anyone has seen this before? Seems unrelated at first glance.

tcclevenger commented:
> In CI checks I am seeing repeatedly:
> 
> ```
> In file included from /home/runner/_work/kokkos/kokkos/kokkos/build/core/unit_test/hip/TestHIPManaged_ViewMapping_a.cpp:2:
> /home/runner/_work/kokkos/kokkos/kokkos/core/unit_test/TestViewMapping_a.hpp:1041:17: error: reference to __host__ function 'use_count' in __host__ __device__ function
>           if (a.use_count() != 3) ++e;
> ```
> 
> Anyone has seen this before? Seems unrelated at first glance.

Odd, it looks like legacy view impl tags `KOKKOS_INLINE_FUNCTION` to use_count
https://github.com/kokkos/kokkos/blob/19b9aff31ec0a98ca46739737b326e02323160a8/core/src/View/Kokkos_ViewLegacy.hpp#L962) 
but new impl doesn't
https://github.com/kokkos/kokkos/blob/19b9aff31ec0a98ca46739737b326e02323160a8/core/src/Kokkos_View.hpp#L1077

But that test is old and the view impl of use_count hasn't been touched for 3 months..

tcclevenger commented:
I tested on Caraway with develop and then this branch, it looks like these changes are to blame. The test that is complaining is launching a parallel_for over a range policy on this space 
https://github.com/kokkos/kokkos/blob/19b9aff31ec0a98ca46739737b326e02323160a8/core/unit_test/TestViewMapping_a.hpp#L1029
and in the kernel calls `a.use_count()` (host only function), which should be fine because it is a host space, but now it complains about being called from a `__host__ __device__` function

janciesko commented:
@tcclevenger Thanks for looking into this. 
The issue is that for `Space = CudaUVMSpace`[*], `Impl::HostMirror<CudaUVMSpace>::Space::execution_space` returns `Kokkos::Cuda`. This follows the logic that `CudaUVMSpace` is host-accessible thus the host mirror keeps the memory space. In this case `Impl::HostMirror<...>::execution_space` is `Kokkos::Serial` but `Impl::HostMirror<...>::memory_space` is `Kokkos::CudaUVMSpace` and `Kokkos::CudaUVMSpace::execution_space` is `Kokkos::Cuda`. 
```
typename Kokkos::Impl::HostMirror<Space>::Space::execution_space; 
```
Given this possible problematic usage, I am inclined towards removing `Space` in `Impl::HostMirror`. 

[*] This applies also to HipManagedSpace and SYCLSharedUSMSpace.

dalg24 commented:
Please resolve conflicts

dalg24 commented on core/unit_test/hip/TestHIP_Spaces.cpp:
```diff
@@ -186,25 +187,32 @@ TEST(hip, space_access) {
       Kokkos::SpaceAccessibility<Kokkos::Impl::HostMirror<Kokkos::HIP>::Space,
                                  Kokkos::HostSpace>::accessible);
 
+#ifndef KOKKOS_IMPL_HIP_UNIFIED_MEMORY
```
Did you mean to commit this?

janciesko replied:
Yes. If `KOKKOS_IMPL_HIP_UNIFIED_MEMORY` is not defined, `HostMirror` is the `HostSpace` memory space. In the opposite case, `HostMirror` remains the`HIPSpace` memory space. I assume this is the behavior we want.

dalg24 commented on core/src/Kokkos_Concepts.hpp:
```diff
@@ -287,6 +287,9 @@ struct is_space {
   static constexpr bool value = is_exe::value || is_mem::value || is_dev::value;
 
+  static constexpr bool is_exec_space() { return is_exe::value; }
+  static constexpr bool is_mem_space() { return is_mem::value; }
```
`is_space<Space>::is_{exec,mem}_space()` would be user-facing and I do not think that is a great addition to the current design

romintomasetti replied:
I agree with @dalg24.

masterleinad replied:
We already have `Kokkos::is_execution_space` and `Kokkos::is_memory_space`.

janciesko replied:
That's right, rolled back change and using those.

crtrott commented on core/src/Kokkos_HostSpace.hpp:
```diff
@@ -144,29 +144,38 @@ static_assert(Kokkos::Impl::MemorySpaceAccess<Kokkos::HostSpace,
-    keep_exe = Kokkos::Impl::MemorySpaceAccess<
-        typename S::execution_space::memory_space,
-        Kokkos::HostSpace>::accessible
+    keep_exe = Kokkos::SpaceAccessibility<S, Kokkos::HostSpace>::accessible
```
Doesn't this mean HostMirror<Device<E,M>> won't work anymore?

janciesko replied:
It should work. Note that `is_space<S>::value` is defined as `value = is_exe::value || is_mem::value || is_dev::value;`

crtrott commented on core/src/Kokkos_HostSpace.hpp:
```diff
@@ -144,29 +144,38 @@ static_assert(Kokkos::Impl::MemorySpaceAccess<Kokkos::HostSpace,
+  // Construct mirror type matching the template parameter type
+  using Space = std::conditional_t<
+      is_space_::is_exec_space(), execution_space,
+      std::conditional_t<is_space_::is_mem_space(), memory_space, Device>>;
```
The meaning of Space has now changed or?

janciesko replied:
The meaning should be unchanged. What changed is that what `Space` is is conditioned on the template parameter `S`. Previously the type of `Space` was conditioned on the accessibility logic.

crtrott commented on core/src/Kokkos_View.hpp:
```diff
@@ -227,8 +227,7 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
       View<typename traits::non_const_data_type, typename traits::array_layout,
-           Device<DefaultHostExecutionSpace,
-                  typename traits::host_mirror_space::memory_space>,
+           typename Impl::HostMirror<memory_space>::Device,
```
Should this be HostMirror<device_type>::Device?

janciesko replied:
Either should work as we construct the same `host_mirror_type` device from both.

crtrott commented on core/src/View/Kokkos_ViewTraits.hpp:
```diff
@@ -418,7 +418,7 @@ struct ViewTraits<std::enable_if_t<Kokkos::is_space<Space>::value>, Space,
   using memory_space    = typename Space::memory_space;
   using host_mirror_space =
-      typename Kokkos::Impl::HostMirror<Space>::Space::memory_space;
+      typename Kokkos::Impl::HostMirror<memory_space>::Space;
```
Why not HostMirror<Device<execution_space, memory_space>>

janciesko replied:
`host_mirror_space` cannot be `HostMirror` but needs to be a `HostMirror::Space`. Also `HostMirror::memory_space` would work, they are just aliases.

ldh4 replied:
I think what he meant might be
`using host_mirror_space = typename Kokkos::Impl::HostMirror<Device<execution_space, memory_space>>::Space;`

janciesko commented on core/unit_test/hip/TestHIP_Spaces.cpp:
```diff
@@ -178,25 +179,32 @@ TEST(hip, space_access) {
   static_assert(Kokkos::SpaceAccessibility<
                 Kokkos::Impl::HostMirror<Kokkos::HIPManagedSpace>::Space,
-                Kokkos::HostSpace>::accessible);
+                Kokkos::HIPManagedSpace>::accessible);
```
@dalg24 Note here - the `HostMirror` memory space to `HIPManagedSpace` is `HIPManagedSpace` because of the accessibility of `HIPManagedSpace` by the host execution space.

janciesko commented:
retest this please

janciesko commented:
Rebased this. We could give this a round of reviews, I hope this is making the use of HostMirror more consistent for the future.

science-enthusiast commented on core/src/Kokkos_HostSpace.hpp:
```diff
@@ -131,29 +131,39 @@ static_assert(Kokkos::Impl::MemorySpaceAccess<Kokkos::HostSpace,
 struct HostMirror {
  private:
+  using is_space_ = typename Kokkos::is_space<S>;
+  static_assert(is_space_::value);
```
It could be `static_assert(typename Kokkos::is_space<S>::value);`

science-enthusiast commented on core/unit_test/TestViewTypedefs.cpp:
```diff
@@ -214,12 +218,12 @@ namespace TestInt {
+  // otherwise if unified memory is on, use DefaultExecutionSpace
+                               std::conditional_t<has_unified_mem_space, Kokkos::DefaultExecutionSpace,
+  // else it's HostSpace
+                                Kokkos::HostSpace>>;
```
Here and in the three other places below, we can shorten the expression:

// If the default exec is a host exec, or if unified memory is on,
// that is it else it's HostSpace
using host_mirror_space = std::conditional_t<is_host_exec || has_unified_mem_space, Kokkos::DefaultExecutionSpace, Kokkos::HostSpace>>;

I have moved the comments together just for my convenience.

janciesko replied:
Good catch, updated.

janciesko committed:
Simplify decision logic for HostSpace

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Simplify decision logic for HostSpace

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko commented:
@science-enthusiast Thanks so much for looking into the this. I have simplified the decision logic.

janciesko commented:
~~@crtrott Could you run a copilot review on this?~~
Never mind, it is still failing,

science-enthusiast commented:
Sure, @janciesko. Let's see whether I can help further with the PR.

Is the Jenkins error with CUDA-12.2-NVCC-RDC again related to CudaUVMSpace? The execution space is being picked up as Cuda instead of Host? All the errors seem to be associated with SIMD and are of the nature "... contains a vector, which is not supported in device code".

Update: The Jenkins CI error is related to the PR on SIMD perf tests merged a little while ago. Other PRs are also facing this error. @tretre91 could say a few words about it here.

tretre91 commented:
> Update: The Jenkins CI error is related to the PR on SIMD perf tests merged a little while ago. Other PRs are also facing this error. @tretre91 could say a few words about it here.

Yes, it seems that on some PR, this job fails because I was using a `View<simd_mask*>` and even though I only use it on host, `View` has `__host__ __device__` methods so it triggers those "... contains a vector, which is not supported in device code" warnings. I opened #8861 to fix this.

tcclevenger commented on core/unit_test/cuda/TestCuda_Spaces.cpp:
```diff
@@ -189,13 +195,20 @@ TEST(cuda, space_access) {
       Kokkos::SpaceAccessibility<Kokkos::Impl::HostMirror<Kokkos::Cuda>::Space,
                                  Kokkos::HostSpace>::accessible);
 
+#ifndef KOKKOS_ENABLE_IMPL_CUDA_UNIFIED_MEMORY
   static_assert(Kokkos::SpaceAccessibility<
                 Kokkos::Impl::HostMirror<Kokkos::CudaSpace>::Space,
                 Kokkos::HostSpace>::accessible);
+#else
+  // Memory space stays the same as host can access CudaSpace
+  static_assert(Kokkos::SpaceAccessibility<
+                Kokkos::Impl::HostMirror<Kokkos::CudaSpace>::Space,
+                Kokkos::CudaSpace>::accessible);
+#endif
```
Is there a reason to start with the negation of KOKKOS_ENABLE_IMPL_CUDA_UNIFIED_MEMORY? I think it is easier to read starting with `#ifdef` instead.

janciesko replied:
I agree, updated.

tcclevenger commented on core/unit_test/cuda/TestCuda_Spaces.cpp:
```diff
@@ -189,13 +195,20 @@ TEST(cuda, space_access) {
 
   static_assert(Kokkos::SpaceAccessibility<
                 Kokkos::Impl::HostMirror<Kokkos::CudaUVMSpace>::Space,
-                Kokkos::HostSpace>::accessible);
```
Is it still worth testing `HostMirror<CudaUVMSpace>` is still accessible by `Kokkos::HostSpace`?

tcclevenger replied:
Similar question for other backends

janciesko replied:
Good point, they're redundant I think from a coverage perspective but maybe worth having from a human readability perspective? No strong opinion. If blocker I remove.

tcclevenger commented on core/src/Kokkos_HostSpace.hpp:
```diff
@@ -131,29 +131,51 @@ static_assert(Kokkos::Impl::MemorySpaceAccess<Kokkos::HostSpace,
+      Kokkos::HostSpace::device_type>;
+
+  using execution_space = typename Device::execution_space;
+  using memory_space    = typename Device::memory_space;
```
Do we need these? I only see them used a single time in the tests.

janciesko commented:
@ndellingwood Could you maybe see if this breaks Trilinos given the changes in behavior? It seems that the CI failures are unrelated.

ndellingwood commented:
> @ndellingwood Could you maybe see if this breaks Trilinos given the changes in behavior? It seems that the CI failures are unrelated.

@janciesko no issues in Trilinos builds, tested Cuda and Hip backends

ldh4 commented on containers/unit_tests/TestDynRankViewTypedefs.cpp:
```diff
@@ -221,12 +221,12 @@ namespace TestInt {
+  // otherwise if unified memory is on, use DefaultExecutionSpace
+                               std::conditional_t<has_unified_mem_space, Kokkos::DefaultExecutionSpace,
+  // else it's HostSpace
+                                  Kokkos::HostSpace>>;
```
Why does this defaults to `HostSpace`, a mem space, instead of `Kokkos::DefaultHostExecutionSpace` or `HostSpace::execution_space`?
Since we can get a mem space from an exec space or vice-versa, it's probably fine, but this doesn't seem consistent.

Also, looking at the changes in ViewTraits.hpp, `ViewTraits::host_mirror_space` now seems to be a mem space. Does that change conflict with the `host_mirror_space` assignment in this test?

janciesko replied:
I think you're right, code updated.

janciesko committed:
Make host_mirror_space more consistent in unit tests

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>
Co-authored-by: Dong Hun Lee <donlee@sandia.gov>

janciesko committed:
Make host_mirror_space more consistent in unit tests

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>
Co-authored-by: Dong Hun Lee <donlee@sandia.gov>

janciesko commented:
@ldh4, re your comment https://github.com/kokkos/kokkos/pull/8252#discussion_r2984833213

```
using host_mirror_space = typename Kokkos::Impl::HostMirror<Device<execution_space, memory_space>>::Space
```
would make`host_mirror_space` a `Device` type. Note that `Space` is always of the same type as the template argument. We could update our tests but it would break downstream codes.

ndellingwood reviewed (approved):
Thanks @janciesko

ndellingwood commented:
OPENACC-NVHPC-CUDA-12.2 [Jenkins failure](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8252/23/stages/?selected-node=186) reported in other PRs (9033):
```
[OPENACC-NVHPC-CUDA-12.2] 5: [ RUN      ] openacc.customscalar_parallel_scan
[OPENACC-NVHPC-CUDA-12.2] 5: /var/jenkins/workspace/Kokkos_PR-8252/core/unit_test/TestCustomScalarParallelScan.hpp:87: Failure
[OPENACC-NVHPC-CUDA-12.2] 5: Expected equality of these values:
[OPENACC-NVHPC-CUDA-12.2] 5:   rowptr_final(i)
[OPENACC-NVHPC-CUDA-12.2] 5:     Which is: 72180
[OPENACC-NVHPC-CUDA-12.2] 5:   prefix_correct(i)
[OPENACC-NVHPC-CUDA-12.2] 5:     Which is: 36096
[OPENACC-NVHPC-CUDA-12.2] 5: [  FAILED  ] openacc.customscalar_parallel_scan (195 ms)
```

CUDA-11.8-Clang-15 [Jenkins failure](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8252/23/stages/?selected-node=364) due to laggy file system issues, unrelated to this PR
```
[CUDA-11.8-Clang-15] [ 51%] Built target Kokkos_CoreUnitTest_CMakePassCmdLineArgswrapper script does not seem to be touching the log file in /var/jenkins/workspace/Kokkos_PR-8252@tmp/durable-ca36e4a1
(JENKINS-48300: if on an extremely laggy filesystem, consider -Dorg.jenkinsci.plugins.durabletask.BourneShellScript.HEARTBEAT_CHECK_INTERVAL=86400)
script returned exit code -1
```

crtrott committed:
Restrict Impl::HostMirror template parameter to memory spaces

Outside of two explicit tests, we only ever instantiated
HostMirror with memory space arguments to begin with.
If we ever reconsider we can change that again, but for now
I think this is a simplification by just requiring that.

Signed-off-by: Christian Trott <crtrott@sandia.gov>

crtrott committed:
Restrict Impl::HostMirror template parameter to memory spaces

Outside of two explicit tests, we only ever instantiated
HostMirror with memory space arguments to begin with.
If we ever reconsider we can change that again, but for now
I think this is a simplification by just requiring that.

Signed-off-by: Christian Trott <crtrott@sandia.gov>

janciesko committed:
Drop unsupported test cases as HostMirror accepts memory spaces only

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Drop unsupported test cases as HostMirror accepts memory spaces only

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

crtrott committed:
Fix issues with HostMirror for MI300A

Signed-off-by: Christian Trott <crtrott@sandia.gov>

crtrott committed:
Clean up SpaceAccssibility and Impl::HostMirror testing

Signed-off-by: Christian Trott <crtrott@sandia.gov>

ndellingwood commented:
SNL-CI HPX job [failed](https://github.com/kokkos/kokkos/actions/runs/24745135402/job/72394742979?pr=8252) to machine running out of space `ccache: error: failed to create temporary file for /home/runner/.cache/ccache/tmp/cpp_stdout.tmp.6U0Ed6.ii: No space left on device`

Jenkins [OPENACC-NVHPC-CUDA-12.2](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8252/28/stages/?selected-node=182) due to laggy filesystem issue,
```
[ 64%] Building CXX object core/unit_test/CMakewrapper script does not seem to be touching the log file in /var/jenkins/workspace/Kokkos_PR-8252@tmp/durable-56d28662
(JENKINS-48300: if on an extremely laggy filesystem, consider -Dorg.jenkinsci.plugins.durabletask.BourneShellScript.HEARTBEAT_CHECK_INTERVAL=86400)
script returned exit code -1
```

Jenkins HIP-ROCm-6.3 [failed](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8252/28/stages/?selected-node=250) due to `fatal error: error in backend: IO failure on output stream: No space left on device` on lascaux_02

janciesko commented:
Retest this please

crtrott commented:
Retest this please.

crtrott commented:
Retest this please!
